kcbq-api/src/main/java/com/wepay/kafka/connect/bigquery/api/TopicAndRecordName.java

```diff
@@ -0,0 +1,116 @@
+package com.wepay.kafka.connect.bigquery.api;
+
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A container class with separate topic and (optional) record names.
+ */
+public class TopicAndRecordName {
+  private final String topic;
+  private final String recordName;
+
+  /**
+   * Convenience method for creating a new {@link TopicAndRecordName} with empty record name.
+   *
+   * @param topic the topic.
+   * @return a new {@link TopicAndRecordName} with empty record name.
+   */
+  public static TopicAndRecordName from(String topic) {
+    return new TopicAndRecordName(topic, null);
+  }
+
+  /**
+   * Convenience method for creating a new {@link TopicAndRecordName}.
+   *
+   * @param topic the topic.
+   * @param recordName the record name.
+   * @return a new {@link TopicAndRecordName}.
+   */
+  public static TopicAndRecordName from(String topic, String recordName) {
+    return new TopicAndRecordName(topic, recordName);
+  }
+
+  /**
+   * Convenience method for creating a new {@link TopicAndRecordName} from {@link SinkRecord}.
+   *
+   * @param record the record used to extract topic and (optionally) record name.
+   * @param includeRecordName whether or not to include the record name.
+   * @return a new {@link TopicAndRecordName}.
+   */
+  public static TopicAndRecordName from(SinkRecord record, boolean includeRecordName) {
+    String maybeRecordName = includeRecordName ? record.valueSchema().name() : null;
+    return from(record.topic(), maybeRecordName);
+  }
+
+  /**
+   * Create a new {@link TopicAndRecordName}.
+   *
+   * @param topic the topic
+   * @param recordName the record name
+   */
+  public TopicAndRecordName(String topic, String recordName) {
+    this.topic = topic;
+    this.recordName = recordName;
+  }
+
+  public String getTopic() {
+    return this.topic;
+  }
+
+  public Optional<String> getRecordName() {
+    return Optional.ofNullable(this.recordName);
+  }
+
+  /**
+   * Convert topic and record name into the schema registry subject.
+   * Subjects follow topic-recordName format.
+   *
+   * If recordName is not present, "value" will be used.
+   *
+   * @return corresponding schema registry subject.
+   */
+  public String toSubject() {
+    return toSubject(KafkaSchemaRecordType.VALUE);
+  }
+
+  /**
+   * Convert topic and record name into the schema registry subject.
+   * Subjects follow topic-recordName format.
+   *
+   * If recordName is not present, "value" or "key" will be used, depending on the schema type.
+   *
+   * @param schemaType schema type used to resolve full subject when recordName is absent.
+   * @return corresponding schema registry subject.
+   */
+  public String toSubject(KafkaSchemaRecordType schemaType) {
```

Comment:
This is leaking Confluent Schema Registry stuff into the API layer, which must be agnostic to specific schema implementations. The concept of a subject is specific to Confluent.

In fact, the entire concept of a record name is specific to the Avro/Confluent implementation. Protobufs will probably work with this since they also have record names, but what about JSON? IIRC, there is another PR that is about supporting JSON schemas.

I am concerned that this PR is far too tied to the Confluent Schema Registry, especially in the API layer.

I am open to changing the API interfaces to make it easier on the Confluent Schema Registry, but not in a way that excludes other potential implementations (especially ones like JSON, where there is active interest). Can you suggest how this can be altered to not leak Confluent Schema Registry assumptions into the API?

Reply:
Good point!
I'm working now on decoupling this plugin from that particular schema implementation. The rough idea is to extract and encapsulate all the SR-related logic in a specialized classes (so that it should be easy to provide support for different schemas).

Reply:
The responsibility for assembling Confluent Schema Registry subject names has been moved to the SchemaRegistrySchemaRetriever.
Also the `recordName` field stays optional, so in case of an alternative implementations (like aforementioned JSON) won't need to bother about it.

Reply:
Ok, so @whynick1 and I have done some talking and investigating. I think we have a good idea about the best approach to achieve not only what you're trying to do, but to evolve KCBQ to support things in a more generic way in general. This amounts to:

1. Adding a pluggable TableRouter that takes a SinkRecord and returns which table it should be written to. (Default: RegexTableRouter)
2. Changing SchemaRetriever interface to have two methods: `Schema getKeySchema(SinkRecord)` and `Schema getValueSchema(SinkRecord)`. (Default: IdentitySchemaRetriever, which just returns `sinkRecord.keySchema()` and `sinkRecord.valueSchema()`)
3. Changing the way that schema updates are handled in the AdaptiveBigQueryWriter.

This approach should give us a ton of flexibility including:

1. It will allow you to pick which table to route each individual message based on all of the information in SinkRecord (topic, key schema, value schema, message payload, etc.)
2. It will allow us to fix some known schema-evolution bugs in cases where one field is added and another is dropped.
3. It will allow us to use SinkRecord's .keySchema and .valueSchema rather than talking to the schema registry for schemas.
4. It will make it easy to support JSON messages, even those that return null for the .keySchema() and .valueSchema() methods--you can implement a custom retriever for this case.

I am going to write up a GH issue today that documents in detail what needs to be done to implement these changes, and we can go from there.

Does this sound good?

Reply:
Add some note regarding @criccomini 's point 1, 2, 3.
1. As an example to your case,
```
public class MultiSchemaRegexSTableRouter {
  Map<TopicAndRecordName, TableId> topicsToBaseTableIds;

  @override
  public String route(SinkRecord sinkRecord) {
    TopicAndRecordName key = new TopicAndRecordName(sinkRecord.topic(), sinkRecord.recordName());
    if (topicsToBaseTableIds.contains(key) {
      return topicsToBaseTableIds.get(key);
    } else {
      // return matched BQ table with topic & record name
    }
  }
}
```
2. Splitting into `getKeySchema`, `valueSchema` also provide an option to load `SinkRecord`'s key schema (if has) to BQ for various purpose including deduplication, debugging, etc.
3. Automatic schema now is default (should be tunable). If turned on, if a new record uses new schema to insert (to BQ), KCBQ will try to update BQ schema from the latest one from Schema Registry. Now, we want to decouple from that, and instead derive new schema from `Record`, assuming json-like `Record` might not has a schema.

Reply:
See https://github.com/wepay/kafka-connect-bigquery/issues/245

Reply:
I really like this idea!

Yesterday I was struggling with changing an existing code on my local branch, responsible for picking up datasets based on both topic and record name, so that it would keep the API part agnostic and completely unaware of different dataset routing strategies. At the end of the day I was a little bit frustrated, because I had to either add yet another, configurable class (similar to SchemaRetriever but for datasets) and letting plugin users to remember about combining the right schema retriever with the right dataset resolver or add responsibility for routing datasets to SchemaRetriever, thus breaking SRP rule...

The proposed `TableRouter.getTable` method solves all the problems!